pennylane_sf/tf.py

```diff
@@ -0,0 +1,275 @@
+            to estimate expectation values of observables. If ``analytic=True``,
+            this setting is ignored.
+        hbar (float): the convention chosen in the canonical commutation
+            relation :math:`[x, p] = i \hbar`
+    """
+    name = "Strawberry Fields TensorFlow PennyLane plugin"
+    short_name = "strawberryfields.tf"
+
+    _capabilities = {
+        "model": "cv",
+        "passthru_interface": "tf",
+    }
+
+    _operation_map = {
+        # Cannot yet support catstates, since they still accept complex parameter
+        # values in Strawberry Fields.
+        # "CatState": Catstate,
+        "CoherentState": Coherent,
+        "FockDensityMatrix": DensityMatrix,
+        "DisplacedSqueezedState": DisplacedSqueezed,
+        "FockState": Fock,
+        "FockStateVector": Ket,
+        "SqueezedState": Squeezed,
+        "ThermalState": Thermal,
+        "GaussianState": Gaussian,
+        "Beamsplitter": BSgate,
+        "CrossKerr": CKgate,
+        "ControlledAddition": CXgate,
+        "ControlledPhase": CZgate,
+        "Displacement": Dgate,
+        "Kerr": Kgate,
+        "QuadraticPhase": Pgate,
+        "Rotation": Rgate,
+        "TwoModeSqueezing": S2gate,
+        "Squeezing": Sgate,
+        "CubicPhase": Vgate,
+        "Interferometer": Interferometer,
+    }
+
+    _observable_map = {
+        "NumberOperator": mean_photon,
+        "TensorN": number_expectation,
+        "X": homodyne(0),
+        "P": homodyne(np.pi / 2),
+        "QuadOperator": homodyne(),
+        "PolyXP": poly_xp,
+        "FockStateProjector": fock_state,
+        "Identity": identity,
+    }
+
+    matrix_gates = {
+        "FockDensityMatrix",
+        "GaussianState",
+        "Interferometer",
+        "FockStateVector",
+    }
+
+    _circuits = {}
+    _asarray = staticmethod(tf.convert_to_tensor)
+
+    def __init__(self, wires, *, cutoff_dim, analytic=True, shots=1000, hbar=2):
+        super().__init__(wires, analytic=analytic, shots=shots, hbar=hbar)
+        self.cutoff = cutoff_dim
+        self.params = dict()
+
+    def apply(self, operation, wires, par):
+        """Apply a quantum operation.
+
+        Args:
+            operation (str): name of the operation
+            wires (Wires): subsystems the operation is applied on
+            par (tuple): parameters for the operation
+        """
+        # convert PennyLane parameter conventions to
+        # Strawberry Fields conventions
+
+        # translate to consecutive wires used by device
+        device_wires = self.map_wires(wires)
+
+        if operation not in self.matrix_gates:
+            # store parameters
+            param_labels = [str(uuid.uuid4()) for _ in range(len(par))]
+
+            for l, v in zip(param_labels, par):
+                self.params[l] = v
+
+            par = self.prog.params(*param_labels)
+
+            if not isinstance(par, Sequence):
+                par = (par,)

```

Comment:
This is because, in the next line, I am doing `*par`, which will fail if `par` is not a sequence 😆 

I could have alternatively done
```python
if isinstance(par, Sequence):
    op = self._operation_map[operation](*par)
else:
    op = self._operation_map[operation](par)
```
but I felt simply always making `par` a sequence was more elegant?

Reply:
Oh, I meant the whole section (from line 214). 😊

Reply:
Oh! Right. Here I am automating the creation of the Strawberry Fields symbolic parameters, which are needed for the TF2 backend to correctly accept tf.Tensor objects:

![image](https://user-images.githubusercontent.com/2959003/90532609-977bec80-e1b6-11ea-9de7-f171fc095226.png)

To name them, I am taking a rather brute-force approach and just using `uuid` to generate hashes (and hoping there are no clashes). Is there a better way of doing this?

Reply:
The `self.prog.params` dictionary simply keeps track of the `{param_label: param_value}` so we can pass them on engine run.

Reply:
I see. Thanks for the explanation! :slightly_smiling_face: